### ISSUE TITLE
🐛 Memory leak with nursery noFloatingPromises

### ISSUE THREAD

**Issue by fgcoelho** [biomejs/biome#7173]
### Environment information

```block
CLI:
  Version:                      2.1.3
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.3.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.14.0

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false

Workspace:
  Open Documents:               0
```

### What happened?

"noFloatingPromises" rule raises my ram usage to 33gb when enabled.

https://github.com/user-attachments/assets/307a4597-01ae-46a1-b937-eed8beae4261

### Expected result

It should not raise my ram usage up to 33gb.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by xKesvaL** [comment]
<img width="961" height="171" alt="Image" src="https://github.com/user-attachments/assets/2f11b21b-1335-4bd2-9d3b-4081f2d2d6be" />

I was wondering why my pc was so slow LOL...

**Comment by dyc3** [comment]
Does it still happen with 2.1.4?

**Comment by xKesvaL** [comment]
> Does it still happen with 2.1.4?

Haven't done extended tests with it yet, just tried it out for a few minutes seems that memory is stable again!

**Comment by fgcoelho** [comment]
Yeah, still happening with 2.1.4. The moment I turn on the rule, ram usage starts going up 1gb/s.

I thought maybe it was a caching issue, so I restarted the dev container, reinstalled node modules, pruned pnpm store, but none of those made any difference.

Wonder why it worked for @xKesvaL 🤔

**Comment by xKesvaL** [comment]
Well for me what's funny is that I wasn't even using that rule. I figured it had to be the same problem, but reading the issue title again it's a nursery rule so has to be turned on manually (I think?).

Weird one for sure.

**Comment by fgcoelho** [comment]
Were you using any version below 2.1.2? They fixed a memory leak in 2.1.2.

**Comment by dyc3** [comment]
Do either of you have biome installed globally? And do the versions differ between the global install and project specific install?

**Comment by fgcoelho** [comment]
No global biome here, and I also run a dev container, so it is isolated. 

*The issue also happens outside the dev container.

**Comment by xKesvaL** [comment]
I was using 2.1.1 globally I think. I'm not sure, since I have it both installed globally and in some projects. Can't seem to reproduce since I have updated my global version to 2.1.4.

**Comment by ematipico** [comment]
Thank you. We need a reproduction to understand what's causing the leak. The issue might come from the code of the user or the code inside the dependencies of the project.

**Comment by fgcoelho** [comment]
No memory leak happened with a new project, I need to clone my current project and exclude stuff until I find the exact cause. Maybe I'll end up opening a new issue.
